### ISSUE TITLE
PowerShell hangs when a WinCompat module import fails due to Core engine modules not being found in $PSHOME\Modules

### ISSUE THREAD

**Issue by Fs00** [PowerShell/PowerShell#13157]
I've encountered an issue using PowerShell **v7.0.1 or higher** in a .NET Windows application.
This bug causes a hang when PS fails to import a module through WinCompat in these two scenarios:
1. the application depends only on System.Management.Automation and Microsoft.PowerShell.Security (the latter is needed because of #14095) but **not on the entire PowerShell SDK**
2. the application depends on the entire PowerShell SDK, but only in self-contained published builds

PS: I know that applications using PowerShell should depend on the entire SDK and not just on System.Management.Automation, but in my case I have an application that only needs to run some commands on the machine and then operate on their output. The full SDK contains 20MB+ of dependencies that my application never uses and significantly bloat the distributable EXE.

## Steps to reproduce

The simplest code that triggers the bug is the following:
```csharp
InitialSessionState sessionState = InitialSessionState.CreateDefault2();
sessionState.ExecutionPolicy = ExecutionPolicy.Unrestricted;  // needed to make import work
using PowerShell pwsh = PowerShell.Create(sessionState);
pwsh.AddScript("Import-Module Dism -UseWindowsPowerShell");
pwsh.Invoke();
Console.WriteLine("Script executed");
```
I have a created a VS project (which can be [downloaded here](https://drive.google.com/file/d/118wAPM5LGOyI3RAdUAxrBN9E_KLjXIpL/view?usp=sharing)) that contains this code and a PublishProfile that can be used to reproduce the bug in the second scenario described above.

**Important:** when removing the project dependency on Microsoft.PowerShell.SDK remember to remove the `bin` directory inside the project folder! Otherwise you'll get inconsistent results.

## Expected behavior

```
Script executed
```
is printed to the standard output and then the application exits gracefully. (in scenario 1 `Import-Module` command will fail anyway, since we don't have `Microsoft.PowerShell.Utility` module that is required for WinCompat)

## Actual behavior

The application hangs indefinitely before printing the `Script executed` message when:
- it depends only on System.Management.Automation and Microsoft.PowerShell.Security v7.0.1+ regardless of the build type
- it depends on Microsoft.PowerShell.SDK v7.0.1+, but **only in Published builds**

### Possible cause
Looking at the [stack trace of the pipeline execution thread](https://pastebin.com/uvQkdxZ0) while the application hangs in scenario 1, the thread appears to be stuck in an infinite loop when trying to import the `Microsoft.PowerShell.Utility` module.
This bug was introduced by #12269. (see my analysis below: https://github.com/PowerShell/PowerShell/issues/13157#issuecomment-741975648)

**Comment by iSazonov** [comment]
I guess the issue could be already fixed in latest builds but you need to replace Appx module in your repro steps with another module because Appx module is not compatible with latest PowerShell 7.1 Preview builds.

**Comment by daxian-dbw** [comment]
@Fs00 The repro project runs correctly for me on Windows 10 1909, where Appx is made compatible with PS 7.0+ (`CompatiblePSEditions = @('Desktop','Core')` in `Appx.psd1`).

What version of Windows are you using? I presume it's not the latest Windows 10 but a downlevel OS, right?
If you are using a downlevel machine, then `Appx` module there may not be compatible with PS 7, and thus the loading of it would go through the [Windows module compatibility feature](https://docs.microsoft.com/en-us/powershell/scripting/whats-new/module-compatibility?view=powershell-7), and it's possible that caused a hang.

Loop in @anmenaga to take a look.

**Comment by Fs00** [comment]
@daxian-dbw I'm using Windows 10 1909 (not the latest cumulative update though, I'm not at home so I cannot check my `Appx.psd1` atm). Did you try it in both the scenarios I listed above?

Anyway, I'm almost sure that on my end my application used WinCompat feature because I happened to see the spawned PowerShell child process in the Task Manager (and now I know why it was there).

**Comment by daxian-dbw** [comment]
> Did you try it in both the scenarios I listed above?

Yes, I was able to run both scenarios successfully.

> Anyway, I'm almost sure that on my end my application used WinCompat feature because I happened to see the spawned PowerShell child process in the Task Manager (and now I know why it was there).

Good to know, so the Appx module might not have the `CompatiblePSEditions = @('Desktop','Core')` in it. @anmenaga is our SME for the WinCompat module feature, and he will be able to help.

**Comment by iSazonov** [comment]
Appx does not work with latest PowerShell 7.1 Preview. See #13138.

**Comment by Fs00** [comment]
@iSazonov Thanks, that's good to know.

**Comment by Fs00** [comment]
@daxian-dbw @anmenaga I've done some more testing and discovered that the bug happens only when importing WinCompat modules.
I have updated the repro project and the issue description to be more accurate. The project now targets PowerShell 7.1 and .NET 5.
I'm confident that you'll now be able to reproduce the issue and have enough information to investigate it. Let me know if there are any problems.

**Comment by anmenaga** [comment]
The `Release.pubxml` has 32-bit `<RuntimeIdentifier>win-x86</RuntimeIdentifier>`. Is this intentional?

**Comment by Fs00** [comment]
That's intentional, although I don't know if that has an impact on the bug (I haven't tested with x64).

**Comment by Fs00** [comment]
Any news on this? Did the latest repro project work?

**Comment by anmenaga** [comment]
The latest repro project did work (when using attached publish profile (`Release.pubxml`), without it it runs fine).
I've looked through WinCompat code and it seems to be behaving as expected, but problems start when it calls PS Remoting code. I assume this has something to do with app running as WoW64 (I vaguely recall remoting bugs with WoW64 from long time ago). This needs more investigation time.

**Comment by Fs00** [comment]
Good to know.
I tried publishing the executable with `<RuntimeIdentifier>win-x64</RuntimeIdentifier>` but it didn't fix the issue, so I don't think the difference in architecture plays a role in this bug.
Did you find the commit that introduced the regression?

**Comment by Fs00** [comment]
Ok @anmenaga, after some research in the codebase I think I got a decent understanding of the issue.
In the stack trace I linked in the first comment you can see that the hang occurs when PowerShell tries to import `Microsoft.PowerShell.Utility` module in order to execute `Export-PSSession`.
Then I saw that PR #12269 introduced a special behavior when WinCompat-ing an engine module:
> If a module falls into one of these buckets, then PS-Core version of the module is imported first **(for engine modules from $PSHOME\Modules)**

The point is, when an application uses only System.Management.Automation (the scenario 1 I mentioned above), the other PS Core engine modules will never be found since they are simply not there. It seems likely that the hang is caused by this edge case not being handled.
At this point my understanding is that before #12269 when a PS Core engine module was not found, it would have been WinCompat-ed. A fix could be to reintroduce this behavior as a fallback, right?

_But what about the scenario 2 (Published build with full SDK), in which all PS Core modules should be there?_
The key to understand this is that **`ModuleIntrinsics.GetPSHomeModulePath()`** ([used here](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/Modules/ImportModuleCommand.cs#L1976)) **returns the path to System.Management.Automation.dll** (returned by Utils.DefaultPowerShellAppBase) **+ '/Modules'** .
The assumption here is that the 'Modules' folder is next to SMA.dll, which is true for Debug builds (`[repro_project_folder]\bin\Debug\net5.0\runtimes\win\lib\net5.0`) but not for Published builds (SMA.dll is near the published EXE, but 'Modules' folder is still inside `runtimes\win\lib\net5.0`).

**Comment by Fs00** [comment]
Hello @anmenaga, is my analysis enough for you to be able to tackle this issue and solve it once and for all?
Unfortunately I don't have the time to look into it by myself, but I'm sure that you would solve it in much less time than I would.
Could it be possible to have this issue fixed in 7.2 stable?
Thanks in advance!

**Comment by Fs00** [comment]
Hello guys.
I'm writing this reply since I've seen no activity on this in the past months.
I think this regression is pretty severe given that it makes impossible to host recent versions of PowerShell (at least in a working way) in self-contained .NET applications. For these scenarios, we need to fallback to PS 7.0.0 which requires .NET Core 3.1 and therefore prevents us from updating to .NET 5+.

For these reasons, I'd like to see at least an expected release for a fix instead of a "Needs-Triage" label after all these months. I feel that this issue is becoming stale.
